include/mbedtls/config.h

```diff
@@ -386,6 +386,20 @@
 //#define MBEDTLS_ECDSA_SIGN_ALT
 //#define MBEDTLS_ECDSA_GENKEY_ALT
 
+/**
+ * \def MBEDTLS_ECDH_LEGACY_CONTEXT
+ *
+ * Use a backward compatible ECDH context. Disabling this makes Mbed TLS use
+ * formally verified implementations for curves where it is available, but in
+ * turn breaks the API.
+ *
+ * \warning     Disabling this option bypasses alternative implementations
+ *              enabled by MBEDTLS_ECDH_GEN_PUBLIC_ALT and
+ *              MBEDTLS_ECDH_COMPUTE_SHARED_ALT for curves which have formally
+ *              verified implementation available.
+ */
+#define MBEDTLS_ECDH_LEGACY_CONTEXT
```

Comment:
Why are we introducing yet another preprocessor symbol? Can we just not deprecate the old interface using the existing methods?

Reply:
Also this is just temporary, this flag would be removed in the next major version.

Another benefit is that until that time we have the freedom to further improve the context and do so in several iterations. With the standard deprecation method this wouldn't be feasible.

We will be adding another primitives and working on other API improvements and I think that the opportunity to use the lessons learned there or just to provide consistency is very well worth temporarily adding a new preprocessor symbol.

Reply:
So this is sort of the opposite of a deprecated interface, it's an interface we don't want on by default yet, but we want to be present alongside the current interface. Historically we've released evaluation code on different branches, not like this, so there's no precedent, which is I think the correct answer to my question.

Therefore we're transitioning from one API to another, and want to retain one while we move to the other. In which case it needs to be clear that the older one will be phased out. It also needs to be clear that this isn't just an 'option' but a stepping stone to changing the API.

Right now, if a user were to look through `config.h` and saw this, it looks like yet another optional feature like all the others. So does it need to be in `config.h`? You can remove the symbol from `config.h` and it'll work just as well, but won't confuse users.

I take it that we want to have the new interface off by default for the moment, and we want to test with the existing the interface, so what advantage are we providing by exposing this to users?

Reply:
"it needs to be clear that the older one will be phased out." -- Should we deprecate the current interface, yet still use the deprecated interface by default?

The argument for having the option in `config.h` seems to be "because all configuration options are in config.h," regardless of whether those options are useful to users. I'm not saying this is a strong argument.

Reply:
The old interface is deprecated. We want people to use the new interface, but we don't want to break the API. Unfortunately we haven't found a way to get the benefits of the new implementation without changing the context structure.

Our intended use case is to add EC implementations that are better at least in some respects than the default implementation, but require a different layout of the context structure. The new context structure allows EC implementations to be added without breaking the API.

We could do a better job of documenting this, but I don't see a better way to encode it.

We do have precedents. For example `MBEDTLS_CMAC_C` defaults off. Not because we don't want people to use CMAC, but because it changes `mbedtls_cipher_context_t`.

Reply:
> The argument for having the option in config.h seems to be "because all configuration options are in config.h," regardless of whether those options are useful to users.

@Patater - That's false. There are examples of configurable options which are local and aren't in `config.h` but are configurable. `config.h` is the public interface of configurable options. There are other options elsewhere such as `MBEDTLS_HAVE_INT32` which aren't in `config.h`.

> The old interface is deprecated. We want people to use the new interface, but we don't want to break the API. 

If the old interface is deprecated, it should get a deprecated warning on build, aside from the issues of how we enable this new interface.